File: JPP/list4/ex2/dh_setup.java
package main;

import java.security.SecureRandom;
import java.util.ArrayList;
import java.util.List;

class DHSetup<T extends GF1234567891> {
    private final T body_generator_;
    private final SecureRandom rand_gen = new SecureRandom();

    public DHSetup(T instance) throws Exception {
        body_generator_ = instance;
        long characteristic = T.Characteristic() - 1;
        List<Long> dividers = new ArrayList<>();
        for(long i = 1; i <= Math.sqrt(characteristic); i++) {
            if(isNumberPrimary(i) && characteristic % i == 0) {
                dividers.add(i);
            }
        }
        do {
            body_generator_.assign(rand_gen.nextLong(characteristic - 1) + 1);
        } while (!isGeneratorLegal(body_generator_, dividers));
    }

    public T getGenerator() {
        return body_generator_;
    }

    public T power(T a, long b) {
        T result = (T) body_generator_.clone();
        result.assign(1);
        T a_copy = (T) a.clone();
        while(b > 0) {
            if(b % 2 == 0) {
                b /= 2;
                a_copy.multiplyEqual(a_copy);
            } else {
                b --;
                result.multiplyEqual(a_copy);
    
                b /= 2;
                a_copy.multiplyEqual(a_copy);
            }
        }
        return result;
    }

    private boolean isNumberPrimary(long number) {
        if(number <= 2) {
            return true;
        }
        for(long i = 2; i < number; i++) {
            if(number % i == 0) {
                return false;
            }
        }
        return true;
    }

    private boolean isGeneratorLegal(T possible_generator, List<Long> dividers) throws Exception {
        long characteristic = T.Characteristic() - 1;
        T one_t = (T) body_generator_.clone();
        one_t.assign(1);
        for(Long val : dividers) {
            T multiplied_val = (T) body_generator_.clone();
            multiplied_val.assign(characteristic / val);
            if(T.equal(T.multiply(possible_generator, multiplied_val), one_t))
                return false;
        }
        return true;
    }
}

File: JPP/list4/ex3/dh_setup/dh_setup.py
from typing import Type, TypeVar, Generic, List
import random
from body.gf import Gf1234577
import copy

T = TypeVar('T', bound=Gf1234577)

class DHSetup(Generic[T]):
    _body_generator: T = None
    _t_class: T = None
    def __init__(self, t_obj: Type[T]) -> None:
        self._t_class = t_obj
        characteristic = self._t_class.Characteristic() - 1
        print(f's {characteristic}')
        dividers = []
        for i in range(1, characteristic ** (1//2) + 1):
            if self.isNumberPrimary(i) and characteristic % i == 0:
                dividers.append(i)
        self._body_generator = self._t_class(random.randint(1, characteristic))
        while not self.isGeneratorLegal(dividers):
            self._body_generator = self._t_class(random.randint(1, characteristic))

    def isNumberPrimary(self, number) -> bool:
        if number <= 2:
            return True
        for i in range(2, number):
            if number % i == 0:
                return False
        return True
    
    def isGeneratorLegal(self, dividers):
        characteristic = self._t_class.Characteristic() - 1
        for val in dividers:
            if self._body_generator * self._t_class(characteristic / val) == self._t_class(1):
                return False
        return True
    
    def getGenerator(self) -> T:
        return self._body_generator
    
    def power(self, a: T, b: int):
        result = self._t_class(1)
        a_copy = copy.deepcopy(a)
        while b > 0:
            if b % 2 == 0:
                b /= 2
                a_copy *= a_copy
            else:
                b-=1
                result *= a_copy

                b /= 2
                a_copy *= a_copy
        return result